### PR TITLE
Version 2.1.0: Add T1135 support, JSON format, and fix year tracking

### LINKED ISSUE #38: Add in T1135 calculation
## T1135 Foreign Income Verification Statement
The T1135 form is a requirement from the Canada Revenue Agency (CRA) for Canadian residents who own specified foreign property with a total cost of more than CAD $100,000 at any time during the tax year. This includes foreign securities held in Canadian brokerage accounts.

### When is T1135 Required?
You must file a T1135 if:
- You are a Canadian resident for tax purposes
- You owned specified foreign property (including foreign stocks) with a total cost exceeding CAD $100,000 at any time during the year
- The securities are held for investment purposes (not personal use)

### PR DESCRIPTION
**Pull request by AC0DEM0NK3Y**
Closes #38 

Version 2.1.0: Add T1135 support, JSON format, and fix year tracking

Major Feature Additions:
- Added T1135 foreign property reporting with new maxcost command
- Added JSON input/output support for transaction data
- Added exchange rate display option for transactions

Bug Fixes:
- Fixed critical bug in year_max tracking where min() was incorrectly used instead of max()

Test Coverage Improvements:
- Added test_year_tracking to verify correct year range calculation
- Added comprehensive test suites for basic transactions and superficial losses
- Added tests for edge cases and complex trading patterns
- Added commission-specific test coverage
- Added JSON format tests
- Improved transaction reader error testing

Code Quality & Documentation:
- Added yapf auto-formatting and fixed all flake8 style issues
- Added comprehensive legal disclaimers and T1135 documentation
- Improved README with better examples
- Added to CONTRIBUTING.md with development guidelines
- Added sample CSV and JSON files for reference

Development Infrastructure:
- Updated tox configuration for multi-Python testing
- Added proper flake8 configuration in pyproject.toml

### CODE REVIEW

**Comment by EmilMaric** [comment]
Thanks for the contributions Andrew and the detailed updates to the README. I'll need to go over these changes throughout the next weeks, but from a quick glance it looks good!

If we're going to update the Python version dependencies in this PR, I think we also need to update the `.travis.yml` file to run the build with those supported Python versions, since right now it's running with Python 3.8 and 3.9 and failing due to the Poetry dependency:
Screenshot 2025-05-19 at 9 14 40 PM

**Comment by EmilMaric** [comment]
Sorry for the delay, I just pushed a change to move the CI/CD from TravisCI to Github Actions, and also updated the supported Python versions as you've done here in this PR. Do you mind resolving the conflicts and re-pushing, and then I'll go through and review the changes?

**Comment by AC0DEM0NK3Y** [comment]
Vacation atm, bit I can look at it in a couple weeks.

**Comment by AC0DEM0NK3Y** [comment]
OK couple weeks was optimistic. I'll have to look at this stuff when I'm not super busy, we have a bit of time until next tax return :)

**Comment by AC0DEM0NK3Y** [comment]
> Sorry for the delay, I just pushed a change to move the CI/CD from TravisCI to Github Actions, and also updated the supported Python versions as you've done here in this PR. Do you mind resolving the conflicts and re-pushing, and then I'll go through and review the changes?

Hi @EmilMaric I brought in your changes and resolved. Tests look good, I also quickly added a small change to warn to user if they are over the $100K threshold for detailed T1135 reporting.

Thanks.

File: capgains/commands/capgains_calc.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 from capgains.ticker_gains import TickerGains
 
 # describes how to align the individual table columns
 colalign = (
-    "left",   # date
-    "left",   # description
-    "left",   # ticker
+    "left",  # date
+    "left",  # description
+    "left",  # ticker
```
Comment (EmilMaric) [inline capgains/commands/capgains_calc.py]: I think this can be reverted, inconsistent indentation with the lines below.

File: tests/helpers.py [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 import stat
+from decimal import Decimal
+
+
+class DecimalEncoder(json.JSONEncoder):
+
+    def default(self, obj):
+        if isinstance(obj, Decimal):
+            return str(obj)
+        return super(DecimalEncoder, self).default(obj)
```
Comment (EmilMaric) [inline tests/helpers.py]: nit: I think this is Python2 style for calling super. You should be able to just call `super().default(obj)`

File: tests/helpers.py [hunk 3]
```diff
@@ -16,22 +26,81 @@ def create_csv_file(directory, filename, data=None, is_readable=True):
...
+                "date": row[0].strftime('%Y-%m-%d')
+                if hasattr(row[0], 'strftime') else row[0],
+                "description": row[1],
+                "ticker": row[2],
+                "action": row[3],
+                "qty": str(row[4]) if isinstance(row[4], Decimal) else row[4],
+                "price": str(row[5])
+                if isinstance(row[5], Decimal) else row[5],
+                "commission": str(row[6])
+                if isinstance(row[6], Decimal) else row[6],
```
Comment (EmilMaric) [inline tests/helpers.py]: I don't think you need the explicitly convert the `Decimal` types to strings here. Your `DecimalEncoder` class already takes care of that when you call `json.dumps(...)`. Dropping that makes the dict look a bit cleaner.

File: tests/helpers.py [hunk 4]
```diff
@@ -16,22 +26,81 @@ def create_csv_file(directory, filename, data=None, is_readable=True):
...
+
+    if not is_readable:
+        open(path, 'a').close()
+        # set all write and execute permissions
+        # to true and read permissions to false (333)
+        os.chmod(
+            path,
+            stat.S_IWUSR | stat.S_IXUSR | stat.S_IWGRP | stat.S_IXGRP
+            | stat.S_IWOTH | stat.S_IXOTH
+        )
```
Comment (EmilMaric) [inline tests/helpers.py]: Perhaps this can be moved out to a helper function in this file as `create_csv_file()` runs the same code to make the file unreadable.

File: tests/test_capgains_maxcost.py [hunk 5]
```diff
@@ -0,0 +1,796 @@
...
+                    t.description,
+                    t.ticker,
+                    t.action,
+                    t.qty,
+                    t.price,
+                    t.commission,
+                    t.currency
+                ]
+            )
+    return path
```
Comment (EmilMaric) [inline tests/test_capgains_maxcost.py]: This file defines its own `create_csv_file()` using `tempfile.mkstemp()` and cleans up with a manual `os.unlink()` in each test. The other test modules (`test_capgains.py`, `test_transactions_reader.py`) use the `testfiles_dir` fixture with the shared `create_csv_file()` in `tests/helpers.py`. Can we switch to that for consistency?
```suggestion
    from tests.helpers import create_csv_file, transactions_to_list

    def test_basic_max_cost(testfiles_dir, requests_mock):
        ...
        csv_path = create_csv_file(
            testfiles_dir, "basic_max_cost.csv", transactions_to_list(transactions)
        )
```

This allows us to remove the 15 `os.unlink()` calls. pytest manages the temp directory, so files can't leak if something raises between creation and cleanup, and the files stay around after a failure so to they can be inspected if needed.

File: tests/test_capgains_maxcost.py [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+                    'd': current_date.isoformat(), 'FXUSDCAD': {
+                        'v': '2.0'
+                    }
+                }
+            )
+            current_date += timedelta(days=1)
+        requests_mock.get(
+            f"{ExchangeRate.valet_obs_url}/FXUSDCAD/json",
+            json={"observations": indicative_observations}
+        )
```
Comment (EmilMaric) [inline tests/test_capgains_maxcost.py]: The Bank of Canada API is currently mocked in four places: the `conftest.py` fixture, `test_capgains_maxcost.py`, `test_comprehensive_basic.py`, `test_comprehensive_superficial.py` and imported by three other test modules.

Suggest replacing them all with one fixture in `conftest.py` that builds observations from the request's `start_date`/`end_date`, skipping weekends like the real Valet API. This should be backwards-compatible with the existing tests as well:

```python
def _valet_callback(series):
    def callback(request, context):
        start = date.fromisoformat(request.qs['start_date'][0])
        end = date.fromisoformat(request.qs['end_date'][0])
        observations = []
        d = start
        while d <= end:
            if d.weekday() < 5:  # Valet only publishes business days
                observations.append(
                    {'d': d.isoformat(), series: {'v': '2.0'}}
                )
            d += timedelta(days=1)
        return {"observations": observations}
    return callback

@pytest.fixture
def exchange_rates_mock(requests_mock):
    for series in ('IEXE0101', 'FXUSDCAD'):
        requests_mock.get(
            f"{ExchangeRate.valet_obs_url}/{series}/json",
            json=_valet_callback(series),
        )
```
